exchanges/okx/okx.go

```diff
@@ -0,0 +1,4465 @@
+		return nil, err
+	}
+	if len(resp) == 1 {
+		return &resp[0], nil
+	}
+	return nil, errNoValidResponseFromServer
+}
+
+// Redeem redemption of investment
+func (ok *Okx) Redeem(ctx context.Context, arg RedeemRequestParam) (*OrderIDResponse, error) {
+	if arg.OrderID == "" {
+		return nil, fmt.Errorf("%w, missing 'orderId'", errMissingRequiredParameter)
+	}
+	if !(arg.ProtocolType == "staking" || arg.ProtocolType == "defi") {
+		return nil, fmt.Errorf("%w, invalid protocol type, only %s and %s allowed", errMissingRequiredParameter, "staking", "defi")
+	}
+	var resp []OrderIDResponse
+	if err := ok.SendHTTPRequest(ctx, exchange.RestSpot, redeemEPL, http.MethodPost, financeRedeem, &arg, &resp, true); err != nil {
+		return nil, err
+	}
+	if len(resp) == 1 {
+		return &resp[0], nil
+	}
+	return nil, errNoValidResponseFromServer
+}
+
+// CancelPurchaseOrRedemption canceld Purchase or Redemptions
+// after cancelling, returning funds will go to the funding account.
+func (ok *Okx) CancelPurchaseOrRedemption(ctx context.Context, arg CancelFundingParam) (*OrderIDResponse, error) {
+	if arg.OrderID == "" {
+		return nil, fmt.Errorf("%w, missing 'orderId'", errMissingRequiredParameter)
+	}
+	if !(arg.ProtocolType == "staking" || arg.ProtocolType == "defi") {
+		return nil, fmt.Errorf("%w, invalid protocol type, only %s and %s allowed", errMissingRequiredParameter, "staking", "defi")
+	}
+	var resp []OrderIDResponse
+	if err := ok.SendHTTPRequest(ctx, exchange.RestSpot, cancelPurchaseOrRedemptionEPL, http.MethodPost, financeCacelPurchase, &arg, &resp, true); err != nil {
+		return nil, err
+	}
+	if len(resp) == 1 {
+		return &resp[0], nil
+	}
+	return nil, errNoValidResponseFromServer
+}
+
+// GetEarnActiveOrders retrives active orders.
+func (ok *Okx) GetEarnActiveOrders(ctx context.Context, productID, protocolType, currency, state string) ([]ActiveFundingOrder, error) {
+	params := url.Values{}
+	if productID != "" {
+		params.Set("productId", productID)
+	}
+	if protocolType == "staking" || protocolType == "defi" {
+		params.Set("protocolType", protocolType)
+	}
+	if currency != "" {
+		params.Set("ccy", currency)
+	}
+	if state == "1" || state == "2" || state == "8" || state == "9" || state == "13" {
+		params.Set("state", state)
+	}
+	var resp []ActiveFundingOrder
+	return resp, ok.SendHTTPRequest(ctx, exchange.RestSpot, getEarnActiveOrdersEPL, http.MethodGet, common.EncodeURLValues(financeActiveOrders, params), nil, &resp, true)
+}
+
+// GetFundingOrderHistory retrives funding order history
+func (ok *Okx) GetFundingOrderHistory(ctx context.Context, productID, protocolType, currency string, after, before time.Time, limit int) ([]ActiveFundingOrder, error) {
+	params := url.Values{}
+	if productID != "" {
+		params.Set("productId", productID)
+	}
+	if protocolType == "staking" || protocolType == "defi" {
+		params.Set("protocolType", protocolType)
```

Comment:
and here protocol type it's optional

Reply:
The main thing I'm trying to address is if someone passes in something else, its not caught. They could set the `protocolType` as "sadklfjasdlfkj" and because you're only adding params when it is a specific string, it can be a successful request, even though the user put in an invalid request.

Reply:
Bump @samuael